mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -0,0 +1,209 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "adjoint"
+
+#include <algorithm>
+#include <iterator>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Errc.h"
+
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace {
+
+template <class T> T isInstanceOf(Operation &op)
+{
+    if (op.getName().getStringRef() == T::getOperationName())
+        return cast<T>(op);
+    else
+        return nullptr;
+}
+
+/// Copy the region of the adjoint operation `op` to the POI specified by the `rewriter`. Build and
+/// return the value mapping `bvm`.
+Value copyAdjointVerbatim(AdjointOp op, PatternRewriter &rewriter, IRMapping &mapping)
+{
+    Block &b = op.getRegion().front();
+    for (auto i = b.begin(); i != b.end(); i++) {
+        if (YieldOp yield = isInstanceOf<YieldOp>(*i)) {
+            assert(++i == b.end() &&
+                   "quantum.yield must be the last operation of an adjoint block");
+            return mapping.lookupOrDefault(yield->getOperand(0));
+        }
+        else {
+            rewriter.insert(i->clone(mapping));
+        }
+    }
+    assert(false && "quantum.yield must present in the adjoint region");
+}
+
+struct AdjointSingleOpRewritePattern : public mlir::OpRewritePattern<AdjointOp> {
+    using mlir::OpRewritePattern<AdjointOp>::OpRewritePattern;
+
+    /// In essence, we build a map from values mentiond in the source data flow to the values of the
+    /// program where quantum control flow is reversed. Most of the time, there is a 1-to-1
+    /// correspondence with a notable exception caused by `insert`/`extract` API asymetry.
+    mlir::LogicalResult matchAndRewrite(AdjointOp adjoint,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        LLVM_DEBUG(dbgs() << "Adjointing the following:\n" << adjoint << "\n");
+        Location loc = adjoint.getLoc();
+        MLIRContext *ctx = adjoint.getContext();
+        assert(adjoint.getRegion().hasOneBlock());
+
+        // First, copy the classical computations directly to the target POI and build the classical
+        // value mapping dictionary.
+        IRMapping classicalMapping;
+        {
+            Block &b = adjoint.getRegion().front();
+            for (auto i = b.begin(); i != b.end(); i++) {
+                if (isa<QuantumDialect>(i->getDialect())) {
+                    /* Skip quantum operations */
+                }
+                else {
+                    LLVM_DEBUG(dbgs() << "classical operation: " << *i << "\n");
+                    rewriter.insert(i->clone(classicalMapping));
+                }
```

Comment:
```suggestion
                if (!isa<QuantumDialect>(i->getDialect())) {
                    LLVM_DEBUG(dbgs() << "classical operation: " << *i << "\n");
                    rewriter.insert(i->clone(classicalMapping));
                }
```

Reply:
or (but I think some people don't like this one.)

```suggestion
                if (isa<QuantumDialect>(i->getDialect())) continue;
                
                 LLVM_DEBUG(dbgs() << "classical operation: " << *i << "\n");
                 rewriter.insert(i->clone(classicalMapping));
```